File: PythonSolution/ShortestPathinBinaryMatrix.py
import math

class Solution:
    def shortestPathBinaryMatrix(self, grid) -> int:
        m = len(grid)
        n = len(grid[0])
        
        if(grid[0][0] == 1 or grid[-1][-1] == 1):
            return -1
        queue = []
        queue.append([0, 0])
        grid[0][0] = 1
        res = 0
        while(len(queue) > 0):
            size = len(queue)
            res = res + 1
            for i in range(size):
                temp = queue.pop(0)
                if(temp[0] == m - 1 and temp[1] == n - 1):
                    return res
                
                for j in range(-1, 2):
                    for k in range(-1, 2):
                        coor = [temp[0] + j, temp[1] + k]
                        if(coor[0] >= 0 and coor[1] >= 0 and coor[0] < m and coor[1] < n and grid[coor[0]][coor[1]] == 0):
                            queue.append(coor)
                            grid[coor[0]][coor[1]] = 1  
        return -1

    def shortestPathBinaryMatrix_(self, grid) -> int:
        n = len(grid)
        if(grid[0][0] or grid[-1][-1]):
            return -1
        q = [(0, 0, 1)]

        # 创建多维数组的方法
        ds = [[math.inf] * n for i in range(n)]

        # for循环迭代
        for i, j, d in q:
            if i == n - 1 and j == n - 1:
                return d
            for m in range(-1, 2):
                for l in range(-1, 2):
                    coor = [m + i, l + j]
                    if(0 <= coor[0] < n and 0 <= coor[1] < n and not grid[coor[0]][coor[1]]):
                        grid[coor[0]][coor[1]] = 1
                        q.append((coor[0], coor[1], d + 1))
        
        return -1

File: Assets/Second/ShortestPathinBinaryMatrix.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class ShortestPathinBinaryMatrix : MonoBehaviour
{
    // Start is called before the first frame update
    void Start()
    {
        
    }

    // Update is called once per frame
    void Update()
    {
        
    }

    /**
    https://leetcode.com/problems/shortest-path-in-binary-matrix/

    BFS，使用DFS会导致TLE
    
     */
    public int ShortestPathBinaryMatrix(int[][] grid) {
        int m = grid.Length;
        int n = grid[0].Length;
        int[][] dir = new int[][]{new int[]{1, 0}, new int[]{0, 1}, new int[]{-1, 0}, new int[]{0, -1}, new int[]{1, 1}, new int[]{1, -1}, new int[]{-1, 1}, new int[]{-1, -1}};
        if(grid[0][0] == 1 || grid[m - 1][n - 1] == 1) return -1;
        
        bool[,] isUse = new bool[m, n];
        
        Queue<int[]> queue = new Queue<int[]>();
        
        queue.Enqueue(new int[]{0, 0});
        int res = 0;
        isUse[0, 0] = true;
        
        while(queue.Count > 0){
            int size = queue.Count;
            res++;
            for(int i = 0; i < size; i++){
                int[] temp = queue.Dequeue();
                
                if(temp[0] == m - 1 && temp[1] == n - 1){
                    return res;
                }
                
                for(int j = 0; j < 8; j++){
                    int x = temp[0] + dir[j][0];
                    int y = temp[1] + dir[j][1];
                    
                    if(x >= 0 && y >= 0 && x < m && y < n && grid[x][y] == 0 && isUse[x, y] == false){
                        queue.Enqueue(new int[]{x, y});
                        isUse[x, y] = true;
                    }
                }
            }
            
        }
        return -1;
    }
}
